File: src/lister.js
/**
 * Port of myindexer to Node
 */ 
var fs = require("fs");
var path = require("path");
var Iconv  = require('iconv').Iconv;
var exec = require('child_process').exec;

function lyrParse(f){
	data = fs.readFile(f, function(err, data){
		conv = new Iconv('TIS-620', 'UTF-8');
		try{
			data = conv.convert(data);
		}catch(e){
		}
		data = data.toString("utf-8");
		data = data.split("\n");
		// in case anyone wants it
		data = {
			"code": f.match(/[/\\]([^\/]+)\.lyr$/i)[1].replace("\\", "/"),
			"name": data[0].replace(/^[\r\n]*|[\r\n]*$/g, ""),
			"artist": data[1].replace(/^[\r\n]*|[\r\n]*$/g, ""),
			"key": data[2].replace(/^[\r\n]*|[\r\n]*$/g, ""),
			"lyric": data.slice(4)
		};
		outstream.write(data.code + "^" + data.name + "^" + data.artist + "^" + data.key + "^" + data.lyric.slice(0,3).join("").replace(/[\r\n]*/g, ""));
	});
}

function scanfolder(folder){
	fs.readdir(folder, function(err, list){
		list.forEach(function(i){
			i = path.join(folder, i);
			stat = fs.statSync(i);
			if(stat.isDirectory()){
				scanfolder(i);
			}else{
				if(Math.random() < 0.01){
					console.log(i);
				}
				try{
					lyrParse(i);
				}catch(e){console.warn("Cannot parse "+i+" : "+e);}
			}
		})
	})
}

if(process.argv.length < 3){
	console.log("node lister.js /path/to/midi/folder");
	process.exit(1);
}
function startScan(){
	scanpath = path.resolve(process.argv[2], "Lyrics");
	fs.readdir(scanpath, function(err, list){
		if(err) throw err;
		list.forEach(function(i){
			p = path.join(scanpath ,i);
			stat = fs.statSync(p);
			if(stat.isDirectory()){
				scanfolder(p);
			}else{
				console.warn("Stray file: "+i);
			}
		});
	});
}

// any less hackish way?
var outpath;
var outstream;
exec('cd ~; pwd', function(err, stdout, stderr){
	if(err) throw err;
	stdout = stdout.replace(/^[\r\n]*|[\r\n]*$/g, "");
	outpath = path.join(stdout, "song.csv");
	fs.writeFileSync(outpath, "");
	outstream = fs.createWriteStream(outpath);
	startScan();
})

File: src/lister.py
import os, re, random, sys

if len(sys.argv) < 2:
	print "lister.py /path/to/midi/folder"
	exit()
outstream = open(os.path.join(os.path.expanduser("~"), "song.csv"), "w");

def lyrParser(f):
	d = open(f).read().decode("tis-620", "ignore").encode("utf-8")
	d=d.split("\n")
	try:
		name = d[0].strip()
	except IndexError:
		name = ""
	try:
		artist = d[1].strip()
	except IndexError:
		artits = ""
	try:
		key = d[2].strip()
	except IndexError:
		key = ""
	try:
		lyric = d[4:]
	except IndexError:
		lyric = []
	return {
		"code": re.findall(r"[/\\](.+?)\.lyr$", f, re.I)[0].replace("\\", "/"),
		"name": name,
		"artist": artist,
		"key": key,
		"lyric": lyric,
	}

def scanfolder(folder):
	for i in os.listdir(folder):
		i = os.path.join(folder, i)
		if os.path.isdir(i):
			scanfolder(i, f)
		else:
			if random.random() < 0.05:
				print i, "\r",
			try:
				lyr = lyrParser(i)
			except Exception, e:
				print "Error with "+i+" "+`e`
				continue
			try:
				lyrics = "".join(lyr['lyric'][1:3])
			except IndexError:
				lyrics = "".join(lyr['lyric'])
			outstream.write("%s^%s^%s^%s^%s\n" % (os.path.abspath(i), lyr['name'], lyr['artist'], lyr['key'], lyrics.replace("\r", "").replace("\n", "")))

for i in os.listdir(os.path.join(os.path.expanduser(sys.argv[1]), "Lyrics")):
	path = os.path.join(os.path.expanduser(sys.argv[1]), "Lyrics", i)
	if not os.path.isdir(path):
		print "Not folder: "+path
		continue
	scanfolder(path)
